blockchain/vm/evm.go

```diff
@@ -538,6 +538,31 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 	return evm.create(caller, codeAndHash, gas, value, contractAddr, humanReadable, codeFormat)
 }
 
+func (evm *EVM) CurrentCodeFormat() params.CodeFormat {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		return params.CodeFormatEVMIstanbulCompatible
+	default:
+		return params.CodeFormatEVM
+	}
+}
+
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// The following code is to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
+		// works after the istanbulIncompatible change
+		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
+		//      (gas price is exactly same as before)
+		if evm.StateDB.GetCodeFormat(caller) == params.CodeFormatEVM {
+			return PrecompiledContractsConstantinople
+		}
+		return PrecompiledContractsIstanbul
+	default:
+		return PrecompiledContractsConstantinople
```

Comment:
Shouldn't this be the recent one?
```suggestion
		return PrecompiledContractsIstanbul
```

Reply:
PrecompiledContractsConstantinople is returned when evm.chainRules.IsIstanbul is false. It means that the IstanbulCompatible block number has not been reached yet. So it returns PrecompiledContractsIstanbul.